### ISSUE TITLE
12/9 進捗報告会11

### ISSUE THREAD

**Issue by PonchTaro** [dp-wsh-lab/b4_s_kato_grad_thesis#16]
### 前回のTODO
* デモ問題をトライしたことが無い問題に対応できるように実装
* できたらpytorchで実装

### 今回の進捗
計画性の無い時間の使い方をした。
##### なぜか先にpytorchを触り出す(しかもpytorch自体を触った時間は短い)
* サーバーの環境を整理してjupyterlabを使えるようにした。
* extensionsを入れて喜んだ
* pytorchの60min_blitzのAutogradまで読んだが、NN以降読めてない

##### デモ問題の考察から逃避行してたかも
* いったん既知の問題で解けたが、他のパラメータで動かない理由を考える
* 行動のインデックスを帰ると結果が学習の成果が変わるので考察
* initとnowを線形ではなく非線形に合わせて値を返してくれる特徴量を考える

**Comment by PonchTaro** [comment]
### TODO
##### 12/12(土)まで
* [ここ](https://colab.research.google.com/drive/13uZd-I1Xx9WVDqovFUZ4apVJAM4HrLy_?usp=sharing#scrollTo=27ipigunrgDo)の分類を3層NNで解く
* MNISTを(0, 1)だけを3層NNで解く。

##### 12/14(月)まで
* ゴールを0で固定した問題をpytorchで線形モデルで解く。-> NNに置き換える

### Agent側の状態表現が生に近い時の行動評価を考える問題
* (0 - 9)までの画像 -> 5, [*2, +1, end, -1, /2] 
* 環境: 状態を数字として知っている。
* Agent: 状態の画像をもっている。画像をただもっているだけで何の数字かは知らない
* 環境の状態が5 -> Agentが+1を選ぶ -> 環境の数字の状態が6に移動 -> 環境から6の画像を手に入れる。

**Comment by PonchTaro** [comment]
# 12/12(土) [実装](https://colab.research.google.com/drive/1E8gs00p5ZEYD-CkMr1BkJCUXC8Jjui_p?usp=sharing)
### Pytorchの学習
基本的に技術の習得は、速くなるように心がけたい。しかし、現実は今は厳しい。
* [60Britz](https://pytorch.org/tutorials/beginner/deep_learning_60min_blitz.html)とDocumentから適当なページを探して読んだ
* MNISTのDataの取り扱いをpytorchとtorchvisionだけで済ませたかったのでDataLoader, Subsetなどの使い方を調べるのに手間取った。
* [結構充実したtutorial](https://pytorch.org/tutorials/)が良いされていたのでこれを読んでもよかった。
* 学習の評価に関しては、full scratchが多かったが、sklearnのmetricsを使った方が短くかけそう。

### 人工データを分類
分類器を学習するということについてもう少し理解したほうが面白いかもしれない。
* [Cross EntropyとKulbuck Divergenceの関係について](https://machinelearningmastery.com/cross-entropy-for-machine-learning/#:~:text=The%20value%20within%20the%20sum,%2B%20KL(P%20%7C%7C%20Q))
* [Binary Cross Entropy](https://towardsdatascience.com/understanding-binary-cross-entropy-log-loss-a-visual-explanation-a3ac6025181a)
*2 Dimensional Standard Distribution から生成した２つのデータに対するNNでの決定境界が直線であることを示す話がここに載っているが、むつかしい。[ベイズ識別方式?](https://www.cs.rit.edu/~rlaz/PatternRecognition/slides/Bayesian.pdf) ->「パターン認識と機械学習」がベイズに偏っているという発言に関係しているかも。

### CNNでの分類に対して
* [LeNet](http://yann.lecun.com/exdb/publis/pdf/lecun-98.pdf)を読んで使ってみた。これじゃなくても全然うまくいく。

**Comment by PonchTaro** [comment]
# Trouble
* tutorialで最適化で用いてる、Cross Entropyで評価するtargetsがone-hot表現になっていないが学習できているのはなぜ。評価時にone-hotにゴニョゴニョしてるのか、自分が思っているcross entropyと違うのか読んで確認。[ドキュメント](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)
* BCEWithLigitsLossだとone-hotにしないと死ぬ。

**Comment by PonchTaro** [comment]
# Trouble ~ 12/14
* OpenGymがwindow systemが存在しないと動かないようだが、jupyter notebookはブラウザ上なので動かず死んだ。virtual Windowを作成してそこで得た結果をmp4に保存したデータをbase64エンコードしてJupyter上にレンダリングする方法を天才くんが考えたみたいなので採用！サーバー上で強化学習とかで作業する人達にとっては普通の発想かも。[天才がまとめたColab](https://colab.research.google.com/drive/1flu31ulJlgiRL1dnN2ir8wGh9p7Zij2t?usp=sharing)
* ただ、pyvirtualdisplayやgymのwrappersを使うだけでは、[こんな](https://github.com/openai/gym/issues/1898)エラーに襲撃を受けたりする。優しい気持ちになったら、原因を特定してみようかと思う。
